File: lib/src/wire.dart
import 'dart:convert';
import 'dart:typed_data';

/// The byte format shared with the Rust side. See `native/src/wire.rs`; the two
/// files describe the same thing and must change together.
///
/// Everything is little-endian and length-prefixed. There is no JSON: keys and
/// ciphertext are already bytes, and re-encoding them would cost more than the
/// format saves.

/// `valueLength` sentinel meaning the row should be deleted rather than written.
const int _tombstone = 0xFFFFFFFF;

/// Which store a record belongs to. Mirrors the `KIND_*` constants in Rust.
enum RecordKind {
  session(1),
  preKey(2),
  signedPreKey(3),
  kyberPreKey(4),
  identity(5);

  const RecordKind(this.code);

  final int code;

  static RecordKind fromCode(int code) => values.firstWhere(
        (kind) => kind.code == code,
        orElse: () => throw WireFormatException('unknown record kind $code'),
      );
}

/// The separator between a name and a device id inside a record key. A control
/// character, so it cannot collide with a UUID or a username. Mirrors
/// `ADDRESS_SEPARATOR` in `native/src/store.rs`.
const String _addressSeparator = '\u0001';

/// Who a session or identity record belongs to.
class SignalAddress {
  const SignalAddress(this.name, this.deviceId);

  final String name;
  final int deviceId;

  @override
  bool operator ==(Object other) =>
      other is SignalAddress && other.name == name && other.deviceId == deviceId;

  @override
  int get hashCode => Object.hash(name, deviceId);

  @override
  String toString() => '$name.$deviceId';
}

/// One row of key material. [value] is null when the row should be deleted —
/// which is how a spent one-time prekey is retired.
class StoredRecord {
  const StoredRecord(this.kind, this.key, this.value);

  final RecordKind kind;
  final String key;
  final Uint8List? value;

  bool get isDeletion => value == null;

  /// The peer this row is about, for session and identity rows. Null for
  /// prekeys, whose key is an id rather than an address.
  SignalAddress? get address {
    if (kind != RecordKind.session && kind != RecordKind.identity) return null;
    final at = key.indexOf(_addressSeparator);
    if (at <= 0) return null;
    final deviceId = int.tryParse(key.substring(at + 1));
    if (deviceId == null) return null;
    return SignalAddress(key.substring(0, at), deviceId);
  }

  @override
  String toString() =>
      'StoredRecord(${kind.name}, $key, ${value == null ? 'deleted' : '${value!.length} bytes'})';
}

class WireFormatException implements Exception {
  const WireFormatException(this.message);

  final String message;

  @override
  String toString() => 'WireFormatException: $message';
}

/// Builds a frame: a count followed by that many fields or records.
class WireWriter {
  final BytesBuilder _body = BytesBuilder(copy: false);
  int _count = 0;

  void field(int tag, List<int> value) {
    _body
      ..add(_u32(tag))
      ..add(_u32(value.length))
      ..add(value);
    _count++;
  }

  void fieldU32(int tag, int value) => field(tag, _u32(value));

  void fieldText(int tag, String value) => field(tag, utf8.encode(value));

  void record(RecordKind kind, String key, List<int>? value) {
    final keyBytes = utf8.encode(key);
    _body
      ..add(_u32(kind.code))
      ..add(_u32(keyBytes.length))
      ..add(keyBytes);
    if (value == null) {
      _body.add(_u32(_tombstone));
    } else {
      _body
        ..add(_u32(value.length))
        ..add(value);
    }
    _count++;
  }

  Uint8List finish() {
    final out = BytesBuilder(copy: false)
      ..add(_u32(_count))
      ..add(_body.takeBytes());
    return out.takeBytes();
  }
}

/// Reads a frame written by Rust.
class WireReader {
  WireReader(this._data) : _view = ByteData.sublistView(_data) {
    if (_data.length < 4) {
      throw const WireFormatException('truncated frame: missing count');
    }
    count = _view.getUint32(0, Endian.little);
  }

  final Uint8List _data;
  final ByteData _view;
  late final int count;
  int _at = 4;

  int _u32() {
    if (_at + 4 > _data.length) {
      throw const WireFormatException('truncated frame: expected a length');
    }
    final value = _view.getUint32(_at, Endian.little);
    _at += 4;
    return value;
  }

  Uint8List _bytes(int length) {
    if (_at + length > _data.length) {
      throw const WireFormatException('truncated frame: value runs past the end');
    }
    // Copy, so the returned slice does not pin the whole frame alive.
    final value = Uint8List.fromList(_data.sublist(_at, _at + length));
    _at += length;
    return value;
  }

  MapEntry<int, Uint8List> field() {
    final tag = _u32();
    return MapEntry(tag, _bytes(_u32()));
  }

  StoredRecord record() {
    final kind = RecordKind.fromCode(_u32());
    final key = utf8.decode(_bytes(_u32()));
    final length = _u32();
    if (length == _tombstone) {
      return StoredRecord(kind, key, null);
    }
    return StoredRecord(kind, key, _bytes(length));
  }

  /// Read the whole frame as fields.
  Map<int, Uint8List> fields() {
    final out = <int, Uint8List>{};
    for (var i = 0; i < count; i++) {
      final entry = field();
      out[entry.key] = entry.value;
    }
    return out;
  }

  /// Read the whole frame as records.
  List<StoredRecord> records() =>
      List.generate(count, (_) => record(), growable: false);
}

Uint8List _u32(int value) =>
    Uint8List(4)..buffer.asByteData().setUint32(0, value, Endian.little);

int readU32(Uint8List value) {
  if (value.length != 4) {
    throw WireFormatException(
      'expected a 4-byte number, got ${value.length} bytes',
    );
  }
  return ByteData.sublistView(value).getUint32(0, Endian.little);
}

File: native/src/wire.rs
//! The byte format shared with the Dart side.
//!
//! Everything crossing the FFI boundary is a flat, little-endian, length-prefixed
//! blob. There is no JSON and no base64: keys and ciphertext are already bytes, and
//! re-encoding them would cost more than the format saves.
//!
//! Two shapes are used:
//!   * a *field list* — `u32 count`, then `u32 tag, u32 len, bytes` per field.
//!     Used for anything with a fixed schema, like a prekey bundle.
//!   * a *record list* — `u32 count`, then `u32 kind, u32 key_len, key,
//!     u32 val_len, val` per record. `val_len == u32::MAX` means "deleted".
//!     Used for everything Dart has to persist.

/// `val_len` sentinel meaning the row should be deleted rather than written.
pub const TOMBSTONE: u32 = u32::MAX;

pub struct Writer {
    buf: Vec<u8>,
    count: u32,
}

impl Writer {
    pub fn new() -> Self {
        // Leave room for the count, which is only known at the end.
        Self {
            buf: vec![0, 0, 0, 0],
            count: 0,
        }
    }

    pub fn field(&mut self, tag: u32, value: &[u8]) {
        self.buf.extend_from_slice(&tag.to_le_bytes());
        self.buf.extend_from_slice(&(value.len() as u32).to_le_bytes());
        self.buf.extend_from_slice(value);
        self.count += 1;
    }

    pub fn field_u32(&mut self, tag: u32, value: u32) {
        self.field(tag, &value.to_le_bytes());
    }

    pub fn record(&mut self, kind: u32, key: &[u8], value: Option<&[u8]>) {
        self.buf.extend_from_slice(&kind.to_le_bytes());
        self.buf.extend_from_slice(&(key.len() as u32).to_le_bytes());
        self.buf.extend_from_slice(key);
        match value {
            Some(value) => {
                self.buf.extend_from_slice(&(value.len() as u32).to_le_bytes());
                self.buf.extend_from_slice(value);
            }
            None => self.buf.extend_from_slice(&TOMBSTONE.to_le_bytes()),
        }
        self.count += 1;
    }

    pub fn finish(mut self) -> Vec<u8> {
        self.buf[..4].copy_from_slice(&self.count.to_le_bytes());
        self.buf
    }
}

pub struct Reader<'a> {
    data: &'a [u8],
    at: usize,
}

impl<'a> Reader<'a> {
    pub fn new(data: &'a [u8]) -> Result<Self, String> {
        if data.len() < 4 {
            return Err("truncated frame: missing count".into());
        }
        Ok(Self { data, at: 4 })
    }

    pub fn count(&self) -> u32 {
        u32::from_le_bytes(self.data[..4].try_into().expect("checked in new"))
    }

    fn u32(&mut self) -> Result<u32, String> {
        let end = self.at + 4;
        if end > self.data.len() {
            return Err("truncated frame: expected a length".into());
        }
        let value = u32::from_le_bytes(self.data[self.at..end].try_into().expect("4 bytes"));
        self.at = end;
        Ok(value)
    }

    fn bytes(&mut self) -> Result<&'a [u8], String> {
        let len = self.u32()? as usize;
        let end = self.at + len;
        if end > self.data.len() {
            return Err("truncated frame: value runs past the end".into());
        }
        let value = &self.data[self.at..end];
        self.at = end;
        Ok(value)
    }

    /// Read the next field as `(tag, value)`.
    pub fn field(&mut self) -> Result<(u32, &'a [u8]), String> {
        let tag = self.u32()?;
        Ok((tag, self.bytes()?))
    }

    /// Read the next record as `(kind, key, value)`, where `None` is a tombstone.
    #[allow(dead_code)] // the record reader lives on the Dart side; kept for symmetry
    pub fn record(&mut self) -> Result<(u32, &'a [u8], Option<&'a [u8]>), String> {
        let kind = self.u32()?;
        let key = self.bytes()?;
        let len = self.u32()?;
        if len == TOMBSTONE {
            return Ok((kind, key, None));
        }
        let end = self.at + len as usize;
        if end > self.data.len() {
            return Err("truncated frame: record value runs past the end".into());
        }
        let value = &self.data[self.at..end];
        self.at = end;
        Ok((kind, key, Some(value)))
    }
}

/// Read a field list into `(tag, value)` pairs.
pub fn fields(data: &[u8]) -> Result<Vec<(u32, &[u8])>, String> {
    let mut reader = Reader::new(data)?;
    let count = reader.count();
    let mut out = Vec::with_capacity(count as usize);
    for _ in 0..count {
        out.push(reader.field()?);
    }
    Ok(out)
}

pub fn u32_field(value: &[u8]) -> Result<u32, String> {
    if value.len() != 4 {
        return Err(format!("expected a 4-byte number, got {} bytes", value.len()));
    }
    Ok(u32::from_le_bytes(value.try_into().expect("4 bytes")))
}
